### PR TITLE
OSSM-1400 Remove support for installing v1.0

### PR DESCRIPTION
**Pull request by dgn**
One of the more impactful changes here is that we no longer default to v1.0 when no version is set. Especially in tests this can lead to new behaviour, but also in conversion- e.g. we were using the same conversion algorithm for status.AppliedSpec conversion, which often had no version (yet), and so would now fail with not defaulting to v1.0.

Leaving this as WIP as I still want to look into the behaviour of this when you have an existing v1.0 control plane and then update to 2.2.0. My plan was for the reconciler to simply log an error and skip reconciliation, but I'm not sure yet if that'll end up spinning the operator into an infinite loop (as the MemberController is also parsing version).

### CODE REVIEW

**Comment by luksa** [comment]
Would it be better to throw an error when the version isn't set instead of defaulting to the latest version?

**Comment by dgn** [comment]
> Would it be better to throw an error when the version isn't set instead of defaulting to the latest version?

That's what happens now. It's just in the conversion that we use DefaultVersion, but that's only to know how to fill some fields, and it only applies to the case where AppliedSpec is empty (and with it spec.Version and would thus fail conversion)

**Comment by jewertow** [comment]
I found in [pkg/apis/maistra/v1/servicemeshcontrolplane_types.go](https://github.com/maistra/istio-operator/blob/maistra-2.2/pkg/apis/maistra/v1/servicemeshcontrolplane_types.go#L112) the following comment:
```
// Existing ServiceMeshControlPlanes with an empty version are treated as
// having the version set to "v1.0"
// +optional
Version string `json:"version,omitempty"`
```
Not very important, but would be nice to update it.

**Comment by jewertow** [comment]
Another comment to update: https://github.com/maistra/istio-operator/blob/maistra-2.2/pkg/controller/servicemesh/webhooks/mutation/controlplane.go#L59.

**Comment by jewertow** [comment]
What about function [ValidateDowngrade](https://github.com/maistra/istio-operator/blob/maistra-2.2/pkg/controller/versions/strategy_v1_1.go#L152) in versionStrategyV1_1? It looks to be no longer needed.

File: pkg/controller/versions/strategy_v1_1.go [hunk 1]
```diff
@@ -215,21 +215,6 @@ func (v *versionStrategyV1_1) ValidateDowngrade(ctx context.Context, cl client.C
 		})
```
Comment (jewertow) [inline pkg/controller/versions/strategy_v1_1.go]: Shouldn't whole this function be removed?

Comment (rcernich) [reply pkg/controller/versions/strategy_v1_1.go]: I believe that's part of the interface, so it needs to be there.  It should return an error, as 1.0 is not supported (i.e. you can't downgrade).

Comment (dgn) [reply pkg/controller/versions/strategy_v1_1.go]: tbh what I never understood about these func is that they're called on downgrade *to* that version and also on downgrade *from* it. So when I'm downgrading 2.1 to 2.0, ValidateDowngrade is called for both, which is really confusing

File: pkg/apis/maistra/conversion/policy.go [hunk 2]
```diff
@@ -203,7 +203,7 @@ func populatePolicyConfig(in *v1.HelmValues, out *v2.ControlPlaneSpec, version v
 		policyType = version.Strategy().GetPolicyType(in, mixerPolicyEnabled, mixerPolicyEnabledSet, remoteEnabled)
 	}
 
-	istiod := !(version == versions.V1_0 || version == versions.V1_1)
+	istiod := !(version == versions.V1_1)
```
Comment (jwendell) [inline pkg/apis/maistra/conversion/policy.go]: nit: `istiod := version != versions.V1_1`

File: pkg/apis/maistra/conversion/policy.go [hunk 3]
```diff
@@ -9,15 +9,15 @@ import (
...
+// between v1.1 and v2.0
 
 func populatePolicyValues(in *v2.ControlPlaneSpec, values map[string]interface{}) error {
 	// Cluster settings
 	if in.Policy == nil {
 		return nil
 	}
 
-	istiod := !(in.Version == "" || in.Version == versions.V1_0.String() || in.Version == versions.V1_1.String())
+	istiod := !(in.Version == "" || in.Version == versions.V1_1.String())
```
Comment (jwendell) [inline pkg/apis/maistra/conversion/policy.go]: nit: `istiod := in.Version != versions.V1_1.String()`

File: pkg/apis/maistra/conversion/security.go [hunk 4]
```diff
@@ -61,7 +61,7 @@ func populateSecurityValues(in *v2.ControlPlaneSpec, values map[string]interface
 			if istiod == nil {
 				break
 			}
-			hasIstiod := !(in.Version == versions.V1_0.String() || in.Version == versions.V1_1.String())
+			hasIstiod := !(in.Version == versions.V1_1.String())
```
Comment (jwendell) [inline pkg/apis/maistra/conversion/security.go]: ditto

File: pkg/apis/maistra/conversion/security.go [hunk 5]
```diff
@@ -115,7 +115,7 @@ func populateSecurityValues(in *v2.ControlPlaneSpec, values map[string]interface
 					break
 				}
 				componentRoot := "pilot"
-				if in.Version == "" || in.Version == versions.V1_0.String() || in.Version == versions.V1_1.String() {
+				if in.Version == "" || in.Version == versions.V1_1.String() {
```
Comment (jwendell) [inline pkg/apis/maistra/conversion/security.go]: ditto

File: pkg/apis/maistra/conversion/telemetry.go [hunk 6]
```diff
@@ -14,7 +14,7 @@ func populateTelemetryValues(in *v2.ControlPlaneSpec, values map[string]interfac
 		return nil
 	}
 
-	istiod := !(in.Version == "" || in.Version == versions.V1_0.String() || in.Version == versions.V1_1.String())
+	istiod := !(in.Version == "" || in.Version == versions.V1_1.String())
```
Comment (jwendell) [inline pkg/apis/maistra/conversion/telemetry.go]: ditto

File: pkg/controller/servicemesh/controlplane/controller_test.go [hunk 7]
```diff
@@ -202,7 +203,9 @@ func newControlPlane() *maistrav2.ServiceMeshControlPlane {
 			Generation: 1,
 			UID:        controlPlaneUID,
 		},
-		Spec:   maistrav2.ControlPlaneSpec{},
+		Spec: maistrav2.ControlPlaneSpec{
+			Version: versions.V2_2.String(),
```
Comment (jwendell) [inline pkg/controller/servicemesh/controlplane/controller_test.go]: perhaps `versions.DefaultVersion`?

File: pkg/controller/servicemesh/member/controller_test.go [hunk 8]
```diff
@@ -460,7 +429,7 @@ func TestReconciliationOfTerminatingNamespace(t *testing.T) {
 				tc.configureMember(member)
 			}
 
-			controlPlane := markControlPlaneReconciled(newControlPlane(""), operatorVersionDefault)
+			controlPlane := markControlPlaneReconciled(newControlPlane("v2.2"), operatorVersionDefault)
```
Comment (jwendell) [inline pkg/controller/servicemesh/member/controller_test.go]: perhaps `versions.DefaultVersion`?

File: pkg/apis/maistra/conversion/policy.go (hunk 2 above)
Comment (dgn) [reply pkg/apis/maistra/conversion/policy.go]: :+1:

File: pkg/apis/maistra/conversion/policy.go (hunk 3 above)
Comment (dgn) [reply pkg/apis/maistra/conversion/policy.go]: not sure about this one. I don't want to break existing 1.0 SMCPs-- just not reconcile them anymore

File: pkg/apis/maistra/conversion/security.go (hunk 4 above)
Comment (dgn) [reply pkg/apis/maistra/conversion/security.go]: :+1:

File: pkg/apis/maistra/conversion/security.go (hunk 5 above)
Comment (dgn) [reply pkg/apis/maistra/conversion/security.go]: see above

File: pkg/controller/servicemesh/controlplane/controller_test.go (hunk 7 above)
Comment (dgn) [reply pkg/controller/servicemesh/controlplane/controller_test.go]: :+1:

File: pkg/controller/servicemesh/member/controller_test.go (hunk 8 above)
Comment (dgn) [reply pkg/controller/servicemesh/member/controller_test.go]: :+1:

File: pkg/apis/maistra/conversion/policy.go (hunk 3 above)
Comment (dgn) [reply pkg/apis/maistra/conversion/policy.go]: actually, I guess the fact that in other places we don't have the check for empty string means that it's fine

File: pkg/controller/versions/strategy_v1_1.go (hunk 1 above)
Comment (dgn) [reply pkg/controller/versions/strategy_v1_1.go]: nevermind, I was mistaking a `<` for a `<=`. This is only called on the downgrade *to* a version.

Comment (dgn) [reply pkg/controller/versions/strategy_v1_1.go]: func is throwing an error now

**Comment by dgn** [comment]
All comments addressed. I just need to find some time to run the upgrade test against an existing v1.0 control plane to be sure

**Review by jwendell — APPROVED** [review]

**Review by jwendell — COMMENTED** [review]
You should remove 1.0 from `build/manifest-templates/clusterserviceversion.yaml`

File: pkg/apis/maistra/conversion/gateways.go [hunk 9]
```diff
@@ -626,8 +626,6 @@ func expansionPortsForVersion(version string) ([]corev1.ServicePort, error) {
 	switch version {
 	case "":
 		fallthrough
-	case versions.V1_0.String():
-		fallthrough
```
Comment (luksa) [inline pkg/apis/maistra/conversion/gateways.go:630]: The "" case now falls through to V1_1. Shouldn't it go to the default version?

Comment (dgn) [reply pkg/apis/maistra/conversion/gateways.go:630]: good catch. I think it shouldn't fall through at all but throw an error. will fix

**Review by luksa — APPROVED** [review]
Everything looks good except the fallthrough.

**Comment by dgn** [comment]
> You should remove 1.0 from `build/manifest-templates/clusterserviceversion.yaml`

thanks! good catch

**Comment by dgn** [comment]
okay nevermind the testing- it is practically impossible these days to install a v1.0 control plane with the 2.1.2 operator. missing CRDs, other errors

**Pull request merged**